team-01/src/main/java/ubc/cosc322/MonteCarloPlayer.java

```diff
@@ -123,38 +134,23 @@ private int playthrough(TreeNode current) {
 	}
 
 	private TreeNode getMaxLeaf(TreeNode root) {
-		ArrayList<TreeNode> leaves = getLeaves(root);
-
-		if (leaves.isEmpty())
-			return root;
-
-		int leafIndex = 0;
-		double maxUCB = leaves.get(0).getUCB();
-		for (int i = 1; i < leaves.size(); i++) {
-			double ucb = leaves.get(i).getUCB();
-			if (ucb > maxUCB) {
-				leafIndex = i;
-				maxUCB = ucb;
-			}
-		}
+		TreeNode node = root;
 
-		return leaves.get(leafIndex);
-	}
+		while (node.children.size() != 0) {
+			double maxUCB = Double.MIN_VALUE;
 
-	private ArrayList<TreeNode> getLeaves(TreeNode root) {
-		ArrayList<TreeNode> leaves = new ArrayList<>();
-
-		for (int i = 0; i < root.children.size(); i++) {
-			TreeNode current = root.children.get(i);
-			// Check if the child has children or is a leaf
-			if (current.children.isEmpty())
-				leaves.add(current);
-			else
-				// Recursively add current's leaves to the list of leaves
-				leaves.addAll(getLeaves(current));
-		}
+			TreeNode nodetoexpand = null;
 
-		return leaves;
+			for (TreeNode child : node.children) {
+				double ucb = child.getUCB();
+				if (ucb > maxUCB) {
+					maxUCB = ucb;
+					nodetoexpand = child;
+				}
+			}
+			node = nodetoexpand;
```

Comment:
This updated `getMaxLeaf()` is significantly more efficient from an algorithmic standpoint, *if it works*. It uses a greedy search to only explore nodes with the highest UCB value at a given depth, potentially skipping a large number of nodes at every iteration, reducing average complexity.

However, this approach operates under the assumption that every node has a UCB value that is greater than or equal to that of any of its children. Have you proven that this assumption is universally true?

Reply:
For what it's worth, I think you are most likely correct in this assumption. Looking over John Levine's video, I believe he uses the same heuristic in his approach.

Reply:
Yes so my understanding is that the parent node should have a UCB that is at least equal to the best node in its children, as when we backpropagate, the wins we backpropagate are added to the parent node resulting in a UCB being greater than or equal to that of any of its children

Reply:
Did you find any evidence to support this? We were struggling to justify the validity of this approach in yesterday's meeting. For example, just because a parent node has the greatest value among its siblings might not necessarily imply that it contains the greatest leaf node as well. For example, this winning parent node may be made of many average children nodes below it, whereas another parent node with slightly less total UCB might have one standout child with the highest actual UCB value of any leaf.

Reply:
I would think that you would want to visit subtrees with higher average UCT rather than visit subtrees with the potential to have 1 high UCT value because a higher average value means it's got better positioning than a node with one really high uct. If a case occurs where the move with the very high UCT is blocked by another move, we would still want to make a move that still results in a good outcome overall.

Reply:
The premise of it seems reasonable. I'm not opposed to doing it that way - it just means we're using a heuristic to prioritize leaves, rather than solely relying on the max UCB. It may end up being stronger to do it this way anyway (and certainly, the function is algorithmically more efficient).